### PR TITLE
Build containers via Job Manager

### PR DESCRIPTION
**Pull request by pditommaso**
This PR implements the container build service via [JobManager](https://github.com/seqeralabs/wave/blob/ac0b877fbc180e0f4cab20c20739198d77296393/src/main/groovy/io/seqera/wave/service/job/JobManager.groovy#L42-L42) and [JobService](https://github.com/seqeralabs/wave/blob/ac0b877fbc180e0f4cab20c20739198d77296393/src/main/groovy/io/seqera/wave/service/job/JobService.groovy#L30-L30)

Main changes: 
* `JobId` class was refactored into [JobSpec](https://github.com/seqeralabs/wave/blob/ac0b877fbc180e0f4cab20c20739198d77296393/src/main/groovy/io/seqera/wave/service/job/JobSpec.groovy#L31-L31) interface, plus `TransferJobSpec`, and `BuildJobSpec` classes 
* JobSpec now defines the max job `maxDuration`
* [JobHandler](https://github.com/seqeralabs/wave/blob/ac0b877fbc180e0f4cab20c20739198d77296393/src/main/groovy/io/seqera/wave/service/job/JobHandler.groovy#L27-L27) has been simplified to single method interface 

@munishchouhan can you please review it and look into tests ? 

 
Solves #575

### CODE REVIEW

**Comment by munishchouhan** [comment]
ok sure, i will work on it

File: src/main/groovy/io/seqera/wave/service/job/JobOperation.groovy [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package io.seqera.wave.service.job
+/**
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ */
+interface JobOperation {
+
+    JobState status(JobSpec jobId)
```
Comment (munishchouhan) [inline src/main/groovy/io/seqera/wave/service/job/JobOperation.groovy:26]: @pditommaso is there a reason variable name is `jobId` or should it be `jobSpec`?

Comment (pditommaso) [reply src/main/groovy/io/seqera/wave/service/job/JobOperation.groovy:26]: Not really. Just a leftover during the refactoring, feel free to update it/them

Comment (munishchouhan) [reply src/main/groovy/io/seqera/wave/service/job/JobOperation.groovy]: done

**Comment by pditommaso** [comment]
Tests are green nice. The most important missing piece here is the use of the job instead of a pod here 

https://github.com/seqeralabs/wave/blob/275fd8acee3fb9ad1412b954f8fc649653ca40c7/src/main/groovy/io/seqera/wave/service/builder/KubeBuildStrategy.groovy#L99-L99

**Comment by munishchouhan** [comment]
> Tests are green nice. The most important missing piece here is the use of the job instead of a pod here
> 
> https://github.com/seqeralabs/wave/blob/275fd8acee3fb9ad1412b954f8fc649653ca40c7/src/main/groovy/io/seqera/wave/service/builder/KubeBuildStrategy.groovy#L99-L99

I will add it

**Comment by munishchouhan** [comment]
I have changed the separator in BuildRequest from '_' to '-'
https://github.com/seqeralabs/wave/blob/6c9a26e024d79407b14620360903e83dae022b2d/src/main/groovy/io/seqera/wave/service/builder/BuildRequest.groovy#L44

**Comment by pditommaso** [comment]
Ouch! better to keep the separator `_` and normalise the `schedulerId` in the Jobfactory 

https://github.com/seqeralabs/wave/blob/275fd8acee3fb9ad1412b954f8fc649653ca40c7/src/main/groovy/io/seqera/wave/service/job/JobFactory.groovy#L44-L56

**Comment by pditommaso** [comment]
@munishchouhan How is the state of this on your side?

**Comment by munishchouhan** [comment]
> @munishchouhan How is the state of this on your side?

Tested locally, its working correctly
```

#13 writing cache image manifest sha256:3211c79449740a2c1d8d187a977cc7e81660666477042e18e1b992e901e359df 0.4s done
#13 DONE 27.7s
------
 > importing cache manifest from docker.io/hrma017/cache:b6266b6b8aead019:
------
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get pods -n wave-local
NAME                             READY   STATUS      RESTARTS   AGE
build-b6266b6b8aead019-1-9skvj   0/1     Completed   0          2m44s
(base) munish.chouhan@Munishs-MacBook-Pro ~ % kubectl get jobs -n wave-local
NAME                       STATUS     COMPLETIONS   DURATION   AGE
build-b6266b6b8aead019-1   Complete   1/1           70s        2m51s
```

**Comment by munishchouhan** [comment]
I will test it in dev now

**Comment by pditommaso** [comment]
It would be nice to add some tests for [Blob job events](https://github.com/seqeralabs/wave/blob/217d89a89356aaf2bb7d4eca43ec2f19244b7c8a/src/main/groovy/io/seqera/wave/service/blob/impl/BlobCacheServiceImpl.groovy#L272-L272) and [Build job events](https://github.com/seqeralabs/wave/blob/217d89a89356aaf2bb7d4eca43ec2f19244b7c8a/src/main/groovy/io/seqera/wave/service/builder/ContainerBuildServiceImpl.groovy#L342-L342)

**Comment by munishchouhan** [comment]
got this error, while testing in dev:
```
2024-09-03 17:45:55	15:45:55.130 [jobs-queue-thread-0] ERROR i.s.w.s.b.ContainerBuildServiceImpl - Unable to build container image 'XXXXXXXXX.dkr.ecr.eu-west-2.amazonaws.com/wave/build/dev:salmon--5ea17099a6004431'; job name=build-5ea17099a6004431-1; cause=Cannot invoke "java.lang.Integer.intValue()" because the return value of "io.seqera.wave.service.job.JobState.getExitCode()" is null
java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "io.seqera.wave.service.job.JobState.getExitCode()" is null
	at io.seqera.wave.service.builder.ContainerBuildServiceImpl.handleJobCompletion(ContainerBuildServiceImpl.groovy:381)
	at io.seqera.wave.service.builder.ContainerBuildServiceImpl.onJobEvent(ContainerBuildServiceImpl.groovy:354)
	at io.seqera.wave.service.job.JobDispatcher.notifyJobCompletion(JobDispatcher.groovy:71)
	at io.seqera.wave.service.job.JobManager.processJob0(JobManager.groovy:88)
	at io.seqera.wave.service.job.JobManager.processJob(JobManager.groovy:68)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
	at io.seqera.wave.service.job.JobManager$_init_lambda1.doCall(JobManager.groovy:63)
	at io.seqera.wave.service.data.stream.AbstractMessageStream.processMessage(AbstractMessageStream.groovy:124)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
	at io.seqera.wave.service.data.stream.AbstractMessageStream$_processMessages_lambda2.doCall(AbstractMessageStream.groovy:138)
	at io.seqera.wave.service.data.stream.impl.RedisMessageStream.consume(RedisMessageStream.groovy:108)
	at io.seqera.wave.service.data.stream.AbstractMessageStream.processMessages(AbstractMessageStream.groovy:138)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:107)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:323)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1268)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1035)
	at org.codehaus.groovy.runtime.InvokerHelper.invokePogoMethod(InvokerHelper.java:1036)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethod(InvokerHelper.java:1019)
	at org.codehaus.groovy.runtime.InvokerHelper.invokeMethodSafe(InvokerHelper.java:97)
	at io.seqera.wave.service.data.stream.AbstractMessageStream$_lambda1.doCall(AbstractMessageStream.groovy:61)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

**Comment by pditommaso** [comment]
The NPE is thrown because [BuildResult](https://github.com/seqeralabs/wave/blob/4087c7889947efdc798636f74afd1f6116241e15/src/main/groovy/io/seqera/wave/service/builder/BuildResult.groovy#L81-L81) uses an `int` for the exitcode instead `JobState` an [Integer](https://github.com/seqeralabs/wave/blob/4087c7889947efdc798636f74afd1f6116241e15/src/main/groovy/io/seqera/wave/service/job/JobState.groovy#L37-L37). 

I've made a quick patch 6f83ad4c. We may improve it later.

**Comment by pditommaso** [comment]
We need to improve in this implementation the handling of the `BuildRequest` object. It needs to be persisted because it can be access in the build completion handler, eg. [here](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/builder/ContainerBuildServiceImpl.groovy#L385-L385). 

Currently is included in the [BuildJobSpec](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/job/spec/BuildJobSpec.groovy#L41-L41), however this does not make much sense from a design point of view and even more it requires too much compute to ser-deserialize the job spec (that happen thousands of times). 

I think it would be better to include it in the object stored via [BuildCacheStore](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/builder/BuildCacheStore.groovy#L43-L43).

Plan: 

1. Add new class `class BuildStoreEntry { BuildRequest request; BuildResult result }` 
2. Using it in place of `BuildResult` in [BuildCacheStore](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/builder/BuildCacheStore.groovy#L43-L43)
3. Modify `BuildJobSpec`to behave more similar to `TransferJobSpec`
4. Move the [schedulerId normalization logic](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/job/spec/BuildJobSpec.groovy#L64-L64) in the [JobFactory](https://github.com/seqeralabs/wave/blob/6f83ad4c099ad35a19cd0e332de137df892218d8/src/main/groovy/io/seqera/wave/service/job/JobFactory.groovy#L39-L39)

@munishchouhan can you give a try to it opening a separate PR?

**Comment by pditommaso** [comment]
There's a failing test, not sure how to address it 

```
should build & push container to azure
Condition not satisfied:

jobService.status(jobSpec).status == JobState.Status.SUCCEEDED
|          |      |        |      |           |
|          |      |        |      false       class io.seqera.wave.service.job.JobState$Status
|          |      |        RUNNING
|          |      io.seqera.wave.service.job.spec.BuildJobSpec(seqeralabs.azurecr.io/wave-tests:97d1c915e9117b48, 2024-09-05T08:44:52.625338442Z, null, build-97d1c915e9117b48-1, seqeralabs.azurecr.io/wave-tests:97d1c915e9117b48, /tmp/test16180240610220047018/97d1c915e9117b48_1, Build)
|          JobState(status:RUNNING, exitCode:null, stdout:null)
<io.seqera.wave.service.job.JobServiceImpl@2dd464ff cleanup=io.seqera.wave.service.cleanup.CleanupStrategy@32589ed7 operations=io.seqera.wave.service.job.impl.DockerJobOperation@6c9dc2b0 transferStrategy=null buildStrategy=io.seqera.wave.service.builder.DockerBuildStrategy@6260a040 jobQueue=io.seqera.wave.service.job.JobQueue@267bf28 jobFactory=io.seqera.wave.service.job.JobFactory@74bf3795>

	at io.seqera.wave.service.builder.ContainerBuildServiceTest.should build & push container to azure(ContainerBuildServiceTest.groovy:241)
```

**Comment by munishchouhan** [comment]
> There's a failing test, not sure how to address it

I will check it out

**Comment by pditommaso** [comment]
The test logic was flawed, it was missing the BuildStore entry. Look at the changes in [78753ed](https://github.com/seqeralabs/wave/pull/622/commits/78753ede22da76f8752c05af2ed0b913af084706). 

Think there's still a failing one

**Comment by munishchouhan** [comment]
> The test logic was flawed, it was missing the BuildStore entry. Look at the changes in [78753ed](https://github.com/seqeralabs/wave/pull/622/commits/78753ede22da76f8752c05af2ed0b913af084706).
> 
> Think there's still a failing one

because BuildStoreEntry get stored before calling the launch function
https://github.com/seqeralabs/wave/blob/c0138115960c7bdf1be2bf969bea6d1d4696289d/src/main/groovy/io/seqera/wave/service/builder/ContainerBuildServiceImpl.groovy#L252

That's why I used jobservice to check the status

**Comment by munishchouhan** [comment]
now launch function is async, so this test needs to use wait for build to finish and then check status in jobservice

**Comment by munishchouhan** [comment]
or we can change from launch to checkOrSubmit

**Comment by munishchouhan** [comment]
> or we can change from launch to checkOrSubmit

but in this case we will get `== Hit build cache for request: `

**Comment by munishchouhan** [comment]
ok got it, you have added the entry for it in the test
nice

**Comment by munishchouhan** [comment]
failed test is redis flaky test, it should pass in the re-

run

**Comment by pditommaso** [comment]
@munishchouhan  Made another round of refactoring on `JobSpec` have a look

File: src/main/groovy/io/seqera/wave/encoder/MoshiEncodeStrategy.groovy [hunk 2]
```diff
@@ -74,6 +74,7 @@ abstract class MoshiEncodeStrategy<V> implements EncodingStrategy<V> {
                         .withSubtype(ProxyHttpResponse.class, ProxyHttpResponse.simpleName)
                         .withSubtype(PairingHeartbeat.class, PairingHeartbeat.simpleName)
                         .withSubtype(PairingResponse.class, PairingResponse.simpleName)
+
```
Comment (pditommaso) [inline src/main/groovy/io/seqera/wave/encoder/MoshiEncodeStrategy.groovy:77]:
```suggestion
```

**Pull request merged**